src/workload.py

```diff
@@ -3,199 +3,83 @@
 # See LICENSE file for licensing details.
 
 """Module containing all business logic related to the workload."""
+import json
 
-from abc import ABC, abstractmethod
-from enum import Enum
-from io import IOBase, StringIO
-
+import ops.pebble
 from ops.model import Container
-from ops.pebble import ExecError
-
-from models import User
-from utils import WithLogging
-
-
-class IOMode(str, Enum):
-    """Class representing the modes to open file resources."""
-
-    READ = "r"
-    WRITE = "w"
-
-
-class AbstractWorkload(ABC):
-    """Abstract class representing general API of a workload, irrespective on the substrate (VM or K8s)."""
-
-    @abstractmethod
-    def start(self):
-        """Execute business-logic for starting the workload."""
-        ...
-
-    @abstractmethod
-    def stop(self):
-        """Execute business-logic for stopping the workload."""
-        ...
-
-    @abstractmethod
-    def health(self) -> bool:
-        """Return the health of the service."""
-        ...
-
-    @abstractmethod
-    def ready(self) -> bool:
-        """Check whether the service is ready to be used."""
-        ...
 
-    @abstractmethod
-    def get_spark_configuration_file(self, mode: IOMode) -> IOBase:
-        """Return the configuration file for Spark History server."""
-        ...
+from common.k8s import K8sWorkload
+from common.utils import WithLogging
+from core.domain import User
+from core.workload import HistoryServerPaths, SparkHistoryWorkloadBase
 
-    @abstractmethod
-    def exec(
-        self, command: str, env: dict[str, str] | None = None, working_dir: str | None = None
-    ) -> str:
-        """Runs a command on the workload substrate."""
-        ...
 
+class SparkHistoryServer(SparkHistoryWorkloadBase, K8sWorkload, WithLogging):
```

Comment:
**question:** Related to my other comment. Why aren't we the implementing methods currently in `K8sWorkload` in `SparkHistoryWorkloadBase`? or better yet, just in plain `SparkHisotryServer`? What does the double inheritance here solve?

Reply:
I noticed that in kafka and zookeeper we are implementing read/write operations in the final class. These are operations very common for which I would believe it is desirable to have common implementations, one for k8s and one for VM. I don't think the business logic there would change between products. Therefore:

* SparkHistoryWorkloadBase is an substrate independent class (where you specify product specific things, e.g. `Paths`, but that are substrate independent)
* K8sWorkload is a product independent class (where you specify substrate specific things, e.g. `read` and `write`)

When you add them together, you can get the most-complete base for the final class, which is substrate-specific and product specific.

Reply:
I believe I created an image at a certain point on Excalidraw.io

![Uploading Screenshot from 2024-03-19 10-57-19.png…]()

Reply:
I think there are two issues with it that I see from first glance.

1. The assumption that any implementation can be 'common'. I don't think that's the case. For charms based on Snaps, we want to `chown` every file on `write()`. That isn't the case for Deb charms. There will be lots of cases where behavior diverges slightly, and we shouldn't be restrictive for that for the sake of it
2. If we DO choose to have a shared implementation for the team, I think it's too early to say that what is presented in this PR is how it will look like. It seems to me like an early optimization to include it here now, and one that increases complexity without any clear upsides

tl;dr - for the current state of the spark charm, I don't think we need double-inheritance. Especially for a charm that only exists on K8s

Reply:
TLDR I don't think using `common` and/or double inheritance creates a lot more complexity (actually provides structure to me), and although I agree it may not be needed by a simple charm (as this one) I see it as a step to foster code-sharing,  common functionalities and collaboration across team. But, here, up to @welpaolo and @theoctober19th . You guys will have to deal with the code more than me, so I don't want to create something that creates friction for you.

Longer reply with my thoughts :smile: 

1. I don't say "ANY" implementation should be in common but only charm-independent ones. We should try to factor out charm-independent functionalities and "read/write"  does seem such a case to me. it is fine to have slightly different behaviour that we can handle with function/constructors arguments, class attributes or different classes (e.g. K8s vs VM). This would provide the foundation for having more homogeneous and maintainable code bases. 

2. Maybe it is not needed in this specific context, but I'm trying to also use these PRs to test/suggest patterns, possibly pushing some ideas out. I wasn't planning to push this into a library and say this is what the code will look like, but I believe it would be good - as a collaboration pattern - if we centralize a `common` module (or another name is fine too :D) with general/common abstractions we see (or also extending existing classes) in **all projects**. It would make discussion easier and provide a first landing space for general functionalities, without needing to release/discuss/agree on general libraries (and be blocked as a feature gets developed) and testing them for some time to evaluate pros/cons. I have implemented this in the past (as a collaboration pattern between different teams with more specific goals), and it worked well

Reply:
I don't have strong reasons to oppose the implementation. However, the only small thing I wanted to ask was, is `SparkHistoryServer extends SparkHistoryWorkloadBase extends K8sWorkload` not a considerable approach to this (since Spark History Server is a K8s only charm)? Basically, what if we use hierarchical inheritance instead of multiple inheritance?

Reply:
I don't have any blocker for the implementation that @deusebio proposed, I like the idea of having a common read/write functions. 
We can evaluate the points raised by @marcoppenheimer in a couple of pulses and revert this approach if it is not working